### ISSUE TITLE
Caller resolver's BadImageFormatException is misreported as a malformed blob

### ISSUE THREAD

**Issue by richlander** [richlander/dotnet-inspect#5759]
`AttributeDecoder.TryDecode`'s public overload accepts a caller-supplied
`Func<string, PrimitiveTypeCode>` resolver for serialized enum names it cannot
resolve locally (`AttributeDecoder.cs:202`, invoked at `:494`). Exceptions the
*caller's* resolver throws are classified by exception type, not by where they
came from, so two of them are reported as "this blob is undecodable."

## Measured

One attribute with a named argument of an unresolvable serialized enum type,
decoded through the resolver overload:

```text
resolver throws BadImageFormatException        -> null (blob reported undecodable)
resolver throws ArgumentOutOfRangeException    -> null (blob reported undecodable)
resolver throws InvalidOperationException      -> propagated InvalidOperationException
resolver throws OutOfMemoryException           -> propagated OutOfMemoryException
```

The blob in the first two rows is well formed. The caller failed, and the
caller is told its input was malformed.

## Contract violated

`docs/design/custom-attribute-value-decoding.md`'s D2 states one rule: an
exception the *blob provokes* is a decode outcome and becomes `null`; an
exception that is *not a statement about the blob* passes through untouched. A
caller resolver's failure is not a statement about the blob — it is the same
category as a caller observer stopping the walk, which D2 already routes through
`MaterializationObserverException` precisely so it is not mistaken for malformed
metadata (#5085).

The resolver boundary has no such tagging, so provenance is lost by the time the
`catch` runs and only the exception type remains.

## Fix direction

Tag the resolver callback boundary the way the observer boundary is tagged: wrap
what the callback throws in a sentinel at the call site, rethrow it past the
malformed-metadata handler, and unwrap at the public edge. This makes D2's rule
decidable at the catch site for both caller boundaries by the same mechanism.

## Consequences for #5732

- The compatibility carve-out currently names #5085 and #5397 as the two
  deliberate behavior corrections; this is a third.
- D2's gate needs a resolver-exception axis: the generator has none today, and
  an unrestricted "`null` wherever SRM throws" oracle would bless the current
  behavior, because SRM observes the callback exception too.

Found by adversarial review of #5732 (round 9).

**Comment by richlander** [comment]
Reconciled under #5288: addressed by merged PR #5815 (merge cde3dab2d4753b3e9c6874b6ed5d21b0e308563d). Resolver callbacks share the provenance sentinel mechanism and are rethrown unchanged at AttributeDecoder. BadImageFormatException and ArgumentOutOfRangeException from the caller are not reported as malformed metadata; the focused resolver provenance cases cover both. The current normative disposition is recorded in docs/design/custom-attribute-value-decoding.md; the slice-4 cleanup migrates the existing gates to the public decode path rather than reopening the repair.
